controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,668 @@
+			log.V(6).Info(fmt.Sprintf("Requeuing after %s as health probe was already run within the last %s",
+				requeueAfter.Truncate(time.Second/10), accessor.config.HealthProbe.Interval))
+			requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+		} else {
+			// Run the health probe
+			tooManyConsecutiveFailures, unauthorizedErrorOccurred := accessor.HealthCheck(ctx)
+			if tooManyConsecutiveFailures || unauthorizedErrorOccurred {
+				// Disconnect if the health probe failed (either with unauthorized or consecutive failures >= HealthProbe.FailureThreshold).
+				accessor.Disconnect(ctx)
+
+				// Store that disconnect was done.
+				didDisconnect = true
+				connected = false //nolint:ineffassign // connected is *currently* not used below, let's update it anyway
+			}
+			switch {
+			case unauthorizedErrorOccurred:
+				// Requeue for connection creation immediately.
+				// If we got an unauthorized error, it could be because the kubeconfig was rotated
+				// and in that case we want to immediately try to create the connection again.
+				log.V(6).Info("Requeuing immediately (disconnected after unauthorized error occurred)")
+				requeueAfterDurations = append(requeueAfterDurations, 1*time.Millisecond)
+			case tooManyConsecutiveFailures:
+				// Requeue for connection creation with the regular ConnectionCreationRetryInterval.
+				log.V(6).Info(fmt.Sprintf("Requeuing after %s (disconnected after consecutive failure threshold met)",
+					accessor.config.ConnectionCreationRetryInterval))
+				requeueAfterDurations = append(requeueAfterDurations, accessor.config.ConnectionCreationRetryInterval)
+			default:
+				// Requeue for next health probe.
+				log.V(6).Info(fmt.Sprintf("Requeuing after %s (health probe succeeded)",
+					accessor.config.HealthProbe.Interval))
+				requeueAfterDurations = append(requeueAfterDurations, accessor.config.HealthProbe.Interval)
+			}
+		}
+	}
+
+	// Send events to cluster sources.
+	lastProbeSuccessTime := accessor.GetLastProbeSuccessTimestamp(ctx)
+	cc.sendEventsToClusterSources(ctx, cluster, time.Now(), lastProbeSuccessTime, didConnect, didDisconnect)
+
+	// Requeue based on requeueAfterDurations (fallback to defaultRequeueAfter).
+	return reconcile.Result{RequeueAfter: minDurationOrDefault(requeueAfterDurations, defaultRequeueAfter)}, nil
+}
+
+// getOrCreateClusterAccessor returns a clusterAccessor and creates it if it doesn't exist already.
+// Note: This intentionally does not already create a client and cache. This is later done
+// via clusterAccessor.Connect() by the ClusterCache reconciler.
+// Note: This method should only be called in the ClusterCache Reconcile method. Otherwise it could happen
+// that a new clusterAccessor is created even after ClusterCache Reconcile deleted the clusterAccessor after
+// Cluster object deletion.
+func (cc *clusterCache) getOrCreateClusterAccessor(_ context.Context, cluster client.ObjectKey) *clusterAccessor {
```

Comment:
Hm yeah let's get rid of it for now. I was iterating a bit on these funcs in general and then I just left the context in case I need it later. But as these are internal funcs we can add the context whenever we need it later